pkg/rbd/nodeserver.go

```diff
@@ -104,16 +99,36 @@ func (ns *NodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 	}
 
 	if !isNotMnt {
-		klog.Infof(util.Log(ctx, "rbd: volume %s is already mounted to %s, skipping"), req.GetVolumeId(), stagingTargetPath)
+		klog.Infof(util.Log(ctx, "rbd: volume %s is already mounted to %s, skipping"), volID, stagingTargetPath)
 		return &csi.NodeStageVolumeResponse{}, nil
 	}
 
+	isLegacyVolume := isLegacyVolumeID(volID)
 	volOptions, err := genVolFromVolumeOptions(ctx, req.GetVolumeContext(), req.GetSecrets(), disableInUseChecks, isLegacyVolume)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
-	volOptions.RbdImageName = volName
-	volOptions.VolID = req.GetVolumeId()
+
+	// get rbd image name from the volume journal
+	// for legacy volumes (v1.0.0), the image name can be found in the staging path
+	if isLegacyVolume {
```

Comment:
can we move this to a separate function? which returns the rbd image name/volID and error?

Reply:
The only reason I refactored this function at all was to reduce cyclomatic complexity. That was already quite high, so any changes of mine triggered it instantly. Apart from that, the changes don't really have something to do with this PR imho. While I applaud cleaning it up further, perhaps dedicated another PR to that.

But if I understand correctly, you'd like to see a function that returns rbdImageName and volID for a given request? I think I can do that.

Reply:
I created  a switch/case for it now, any further improvements can be done outside of this PR imho